dds/DCPS/SubscriberImpl.cpp

```diff
@@ -569,15 +571,17 @@ SubscriberImpl::get_datareaders(
 DDS::ReturnCode_t
 SubscriberImpl::notify_datareaders()
 {
-  ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex,
-                   guard,
-                   this->si_lock_,
-                   DDS::RETCODE_ERROR);
-
+  DataReaderMap localreadermap;
   DataReaderMap::iterator it;
-
-  for (it = datareader_map_.begin(); it != datareader_map_.end(); ++it) {
-    if (it->second->have_sample_states(DDS::NOT_READ_SAMPLE_STATE)) {
+  {
+    ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex,
+                    guard,
+                    this->si_lock_,
+                    DDS::RETCODE_ERROR);
+    localreadermap = datareader_map_;
+  }
+  for (it = localreadermap.begin(); it != localreadermap.end(); ++it) {
```

Comment:
Does the `it` object need to outlive the for loop?  I realize it was that way before this PR, but it would be good to reduce its scope to the smallest necessary.

Reply:
My first inclination was to minimize changes, but I agree that the iterator is not needed again outside the loop.